lib/sshutils/x11/forward.go

```
@@ -0,0 +1,229 @@
+	}
+
+	ok, err := sess.SendRequest(sshutils.X11ForwardRequest, true, ssh.Marshal(payload))
+	if err != nil {
+		return trace.Wrap(err)
+	} else if !ok {
+		return trace.Errorf("x11 forward request failed")
+	}
+
+	return nil
+}
+
+type x11ChannelHandler func(ctx context.Context, nch ssh.NewChannel)
+
+// ServeX11ChannelRequests opens an x11 channel handler and starts a
+// goroutine to serve any channels received with the handler provided.
+func ServeX11ChannelRequests(ctx context.Context, clt *ssh.Client, handler x11ChannelHandler) error {
+	channels := clt.HandleChannelOpen(sshutils.X11ChannelRequest)
+	if channels == nil {
+		return trace.AlreadyExists("x11 forwarding channel already open")
+	}
+
+	go func() {
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+		for {
+			select {
+			case nch := <-channels:
+				if nch == nil {
+					return
+				}
+				go handler(ctx, nch)
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	return nil
+}
+
+// OpenNewXServerListener opens a new XServer unix socket with the first unused display
+// number. The unix socket's corresponding XServer display value will also be returned.
+func OpenNewXServerListener(displayOffset int, screen uint32) (net.Listener, string, error) {
+	for displayNumber := displayOffset; displayNumber < displayOffset+x11MaxDisplays; displayNumber++ {
+		if l, err := net.Listen("unix", xserverUnixSocket(displayNumber)); err == nil {
+			return l, unixXDisplay(displayNumber), nil
+		}
+	}
+
+	return nil, "", trace.LimitExceeded("No more x11 sockets are available")
+}
+
+// Forward begins x11 forwarding between an xserver connection and an x11 channel.
+// The xserver connection may be a direct xserver connection, or another x11 channel.
+func Forward(xconn io.ReadWriteCloser, x11Chan ssh.Channel) {
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		io.Copy(x11Chan, xconn)
+		x11Chan.CloseWrite()
+	}()
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		io.Copy(xconn, x11Chan)
```

The same as above. I don't think we should ignore errors here?